cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt

```diff
@@ -129,10 +141,10 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
      * a [CollectionComprehension].
      */
     private fun handleSetComprehension(node: Python.AST.SetComp): CollectionComprehension {
-        return newCollectionComprehension(rawNode = node).apply {
+        return newCollectionComprehension(rawNode = node).applyWithScope {
```

Comment:
I am not sure, if this is really relevant in our case, but it seems that the exact definition is that only the control variable is not leaked into the outer scope, but I guess the only sane way to model this is exactly like you did by opening up a local scope for the comprehension.

Reply:
Interesting, from https://peps.python.org/pep-0709/ -> "Comprehensions are currently compiled as nested functions, which provides isolation of the comprehension’s iteration variable, but is inefficient at runtime."

Reply:
> Hm, this sounds as if the current behavior would be appropriately modeled with the local scope but the intended/future behavior might be different.

Yes, although I am not sure if this "new" bevahiour will leak the control variable or not. Maybe our PEP master @maximiliankaul understands this better than us.

> 
> Actually, we're only generating a `VariableDeclaration` inside this local scope for the variable of the comprehension expression and this is done inside the `PythonAddDeclarationsPass`. Leaving the scope earlier (i.e., right after generating the `ComprehensionExpression`) here won't impact on resolving variables, right?

I would assume so

Reply:
From my understanding, the variable should be available only in the `LocalScope` tied to the `ComprehensionExpression`. Obviously, we can always enter this scope to incorrectly resolve a variable but this shouldn't be the case. I also added a test which simulates the test case he provided in the original comment.

Reply:
I added a switch for different versions. For all versions > python 3.0.0, we create the `LocalScope`. If nothing is specified or for smaller versions, we do not create the `LocalScope`. I also added a test for it